beacon_node/network/src/service.rs

```diff
@@ -943,6 +949,49 @@ impl<T: BeaconChainTypes> NetworkService<T> {
         // backfill sync to start over from it. Then make backfill sync use a higher CGC (say 128)
         // and when oldest_block is less than the oldest step with a value < 128 we can delete that
         // step such that `custody_group_count(clock - da_window)` returns 128.
+        //
+        //                    CGC 128
+        //             :      ___________:____________
+        //             :     |           :
+        // CGC 8       :     |           :
+        // ............:_____|           :
+        //             : WS start        : Finalized
+        //     <------   block             block
+        //  backfill
+        //  with CGC 8
+        //
+        //
+        //
+        //                    CGC 128
+        // .............................. ___________:____________
+        //                               :
+        //                               :
+        //                               :
+        //                               : Finalized
+        //                       <------   block
+        //                   backfill
+        //                   with CGC 128
+        //
+        let oldest_block_slot = self.beacon_chain.store.get_anchor_info().oldest_block_slot;
+        let finalized_slot = self.beacon_chain.finalized_slot();
+        let cgc_at_oldest_block_slot = self.network_globals.custody_group_count(oldest_block_slot);
+        let cgc_at_finalized_slot = self.network_globals.custody_group_count(finalized_slot);
+        // TODO(das): If we support a decreasing CGC we must consider the min value between this two
+        // slots.
+        if cgc_at_oldest_block_slot < cgc_at_finalized_slot {
+            let backfill_started_recently = finalized_slot.saturating_sub(oldest_block_slot)
+                < MAX_SLOT_DISTANCE_BACKFILL_RESTART;
+            // Note: we don't check if backfill finished. If it did because we are close to genesis,
+            // we want to restart it anyway to backfill with the CGC. The only condition to NOT
+            // restart is if backfill went too far and thus we would waste too much bandwidth
+            // fetching the blocks again.
+            if backfill_started_recently {
+                // We need backfill sync to fetch batches with `CGC_f = cgc_at_finalized_slot`. Then
+                // `custody_group_count(oldest_block_slot) should now return `CGC_f`. So we have to
+                // delete the CGC updates with `update.slot < finalized_slot`
+                todo!();
```

Comment:
To prevent concurrent updates of `oldest_block_slot`, we can get `import_historical_block_batch` to handle the mutation for us.

We can add an argument to that function, something like `restart: bool`, and set the `oldest_block_{slot,root}` to the _last_ block in the batch passed in (allowing the `oldest_block_slot` to _go up_).

TODO: what if state reconstruction is running as well?